carp.deployment.core/src/commonMain/kotlin/dk/cachet/carp/deployment/application/DeploymentService.kt

```diff
@@ -18,7 +17,7 @@ interface DeploymentService
     /**
      * Instantiate a study deployment for a given [StudyProtocolSnapshot].
      *
-     * @throws InvalidConfigurationError when [protocol] is invalid.
+     * @throws IllegalArgumentException when [protocol] is invalid.
```

Comment:
We should really consider still having a hierarchy of exceptions pertaining to the domain. It is easier to handle these in a special way if needed in the implementations and they can be enriched with information and behavior specific to the domain unlike the generic unchecked exceptions

Reply:
Agreed we need to talk about this, but `InvalidConfigurationError` was mostly used haphazardly, so for now opted to simply have `IllegalArgumentException` uniformly for anywhere the _arguments_ are incorrectly.

Broadly speaking every `IllegalArgumentException` that comes out of services right now should safely map to a 400 of corresponding web requests. The problem lies in when the underlying _domain objects_ throw `IllegalArgumentException`'s which do not map to illegal arguments of the service requests.

For that reason it would make sense to have domain specific `IllegalArgumentException`s for application services, so that we can uniformly map domain-specific exceptions to web errors. In fact, we talked about this before in the past. I will add this to the `tech-meeting` backlog and see what previous information I can gather.

Reply:
https://app.zenhub.com/workspaces/mobile-tech-meetings-5fcf567a99061f00110291ae/issues/imotions/mobile-project-management/26